File: native/EasyVirtualDisplay.Vdd/Domain/Contracts.cs
using System;
using System.Collections.Generic;

namespace EasyVirtualDisplay.Vdd.Domain;

public static class DriverStatuses
{
    public const string Ok = "ok";
    public const string Inaccessible = "inaccessible";
    public const string Unknown = "unknown";
    public const string UnknownProblem = "unknown_problem";
    public const string Disabled = "disabled";
    public const string DriverError = "driver_error";
    public const string RestartRequired = "restart_required";
    public const string DisabledService = "disabled_service";
    public const string NotInstalled = "not_installed";
}

public enum DriverState
{
    Ok,
    Inaccessible,
    Unknown,
    UnknownProblem,
    Disabled,
    DriverError,
    RestartRequired,
    DisabledService,
    NotInstalled,
}

public static class ParentGpus
{
    public const string Auto = "auto";
    public const string Nvidia = "nvidia";
    public const string Amd = "amd";
}

public static class Orientations
{
    public const string Landscape = "landscape";
    public const string Portrait = "portrait";
    public const string LandscapeFlipped = "landscape_flipped";
    public const string PortraitFlipped = "portrait_flipped";
}

public sealed class DisplayMode
{
    public int Width { get; init; }

    public int Height { get; init; }

    public int Hz { get; init; }
}

public sealed class SupportedResolution
{
    public int Width { get; init; }

    public int Height { get; init; }

    public IReadOnlyList<int> RefreshRates { get; init; } = Array.Empty<int>();
}

public sealed class DisplaySummary
{
    public int Index { get; init; }

    public int Identifier { get; init; }

    public string DeviceName { get; init; } = string.Empty;

    public string DisplayName { get; init; } = string.Empty;

    public bool Active { get; init; }

    public DisplayMode? CurrentMode { get; init; }

    public string CurrentOrientation { get; init; } = Orientations.Landscape;

    public IReadOnlyList<SupportedResolution> SupportedResolutions { get; init; } = Array.Empty<SupportedResolution>();

    public bool UnsupportedCurrentMode { get; init; }
}

public sealed class HostSnapshot
{
    public long Revision { get; init; }

    public string Status { get; init; } = DriverStatuses.Unknown;

    public string DriverVersion { get; init; } = "0.0.0.0";

    public int MaxDisplays { get; init; }

    public IReadOnlyList<DisplaySummary> Displays { get; init; } = Array.Empty<DisplaySummary>();

    public IReadOnlyList<DisplayMode> CustomModes { get; init; } = Array.Empty<DisplayMode>();

    public string ParentGpu { get; init; } = ParentGpus.Auto;
}

public sealed class SetDisplayModeInput
{
    public int Index { get; init; }

    public int? Width { get; init; }

    public int? Height { get; init; }

    public int? Hz { get; init; }

    public string? Orientation { get; init; }
}

public sealed class ApplyAdminConfigInput
{
    public IReadOnlyList<DisplayMode> CustomModes { get; init; } = Array.Empty<DisplayMode>();

    public string ParentGpu { get; init; } = ParentGpus.Auto;
}

public sealed class AdminConfigSnapshot
{
    public IReadOnlyList<DisplayMode> CustomModes { get; init; } = Array.Empty<DisplayMode>();

    public string ParentGpu { get; init; } = ParentGpus.Auto;
}

File: src/shared/contracts.ts
export const DRIVER_STATUSES = [
  'ok',
  'inaccessible',
  'unknown',
  'unknown_problem',
  'disabled',
  'driver_error',
  'restart_required',
  'disabled_service',
  'not_installed'
] as const

export type DriverStatus = (typeof DRIVER_STATUSES)[number]

export const PARENT_GPUS = ['auto', 'nvidia', 'amd'] as const

export type ParentGpu = (typeof PARENT_GPUS)[number]

export const ORIENTATIONS = [
  'landscape',
  'portrait',
  'landscape_flipped',
  'portrait_flipped'
] as const

export type Orientation = (typeof ORIENTATIONS)[number]

export const APP_THEMES = ['system', 'light', 'dark'] as const

export type AppTheme = (typeof APP_THEMES)[number]

export const APP_LANGUAGES = ['system', 'en', 'zh-CN'] as const
export type AppLanguage = (typeof APP_LANGUAGES)[number]
export type EffectiveLanguage = 'en' | 'zh-CN'

export interface DisplayMode {
  width: number
  height: number
  hz: number
}

export interface SupportedResolution {
  width: number
  height: number
  refreshRates: number[]
}

export interface DisplaySummary {
  index: number
  identifier: number
  deviceName: string
  displayName: string
  active: boolean
  currentMode: DisplayMode | null
  currentOrientation: Orientation
  supportedResolutions: SupportedResolution[]
  unsupportedCurrentMode: boolean
}

export interface HostSnapshot {
  revision: number
  status: DriverStatus
  driverVersion: string
  maxDisplays: number
  displays: DisplaySummary[]
  customModes: DisplayMode[]
  parentGpu: ParentGpu
}

export interface AppSettings {
  launchOnLogin: boolean
  closeToTray: boolean
  startMinimized: boolean
  fallbackDisplay: boolean
  keepScreenOn: boolean
  theme: AppTheme
  language: AppLanguage
}

export interface AppSnapshot {
  host: HostSnapshot
  settings: AppSettings
  effectiveLanguage: EffectiveLanguage
}

export const DEFAULT_APP_SETTINGS: AppSettings = {
  launchOnLogin: false,
  closeToTray: true,
  startMinimized: false,
  fallbackDisplay: false,
  keepScreenOn: false,
  theme: 'system',
  language: 'system'
}

export const EMPTY_HOST_SNAPSHOT: HostSnapshot = {
  revision: 0,
  status: 'unknown',
  driverVersion: 'pending',
  maxDisplays: 0,
  displays: [],
  customModes: [],
  parentGpu: 'auto'
}

export const EMPTY_APP_SNAPSHOT: AppSnapshot = {
  host: EMPTY_HOST_SNAPSHOT,
  settings: DEFAULT_APP_SETTINGS,
  effectiveLanguage: 'en'
}
